proposals/redirects.md

```diff
@@ -0,0 +1,116 @@
+<!--
+  Note: You are probably looking for `stage-1--discussion-template.md`!
+  This template is reserved for anyone championing an already-approved proposal.
+
+  Community members who would like to propose an idea or feature should begin
+  by creating a GitHub Discussion. See the repo README.md for more info.
+
+  To use this template: create a new, empty file in the repo under `proposals/${ID}.md`.
+  Replace `${ID}` with the official accepted proposal ID, found in the GitHub Issue
+  of the accepted proposal.
+-->
+
+- Start Date: 2023-05-22
+- Reference Issues: https://github.com/withastro/roadmap/issues/466
+- Implementation PR: https://github.com/withastro/astro/pull/7067/
+
+# Summary
+
+Add a `redirects` config option to the Astro config which allows you to define redirects in a central place. Allow integrations to read this information and apply it to it's own redirects configuration for deployment.
+
+# Example
+
+New config option used like so:
+
+```js
+import { defineConfig } from 'astro/config';
+
+export default defineConfig({
+  redirects: {
+    '/other': '/place'
+  }
+});
+```
+
+# Background & Motivation
+
+This was original proposed as a stage 1 discussion [here](https://github.com/withastro/roadmap/discussions/319) and was one of the top voted proposals.
+
+As websites age there are times where routes are rearchitectured. In order preserve existing links to content on the web it is common to set up redirects in your web server.
+
+Redirect configuration varies depending on what web server or host you use. Some times you might even want to change hosts, in which case your redirects need to be converted over to a new format.
+
+Having a redirects configuration within Astro itself allows a single place to define redirects that works everywhere.
+# Goals
+
+- Works in development mode.
+- Writes out `<meta http-equiv="refresh">` tags in static builds.
+- Gives proper hooks to integrations so that they can write to their own configuration.
+
+# Non-Goals
+
+- Specifying temporary redirects. These are best done at runtime, so SSR is a more appropriate way to do those.
+    - In the future we could expand this API to give more control over what status code is used, once use-cases are known.
```

Comment:
[3xx](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#redirection_messages) supports different sorts of redirects, so I'm not sure if we should limit this as a start. I think the usecase is common that some status codes makes more sense for their business logic.

Maybe for SSG we only support a subset of status codes?

Reply:
That's fair. The way we would probably do that is:

```js
{
  '/one': {
    status: 302,
    destination: '/two'
  }
}
```

TypeScript and runtime would restrict to 3xx statuses.

Are you ok with keeping the string destination as a shortcut for 301 (which is the more common one for the use-cases).

Reply:
Yeah I'm fine with a `string` shortcut since we won't be merging the settings (which would be complex otherwise).

Reply:
Updated to allow the object syntax and to show dynamic routes.